### PR TITLE
Authenticating SAML user by UserDetailService

### PR DESCRIPTION
**Pull request by herrminni**
Added support for authenticating a SAML user by UserDetailsService.
This enhances the AuthenticationProvider by adding UserDetail metadata
to the principal and also set authorities like roles.

Closes gh-7465

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@herrminni Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=7550)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-security&pullRequestId=7550) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@herrminni Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-security&pullRequestId=7550)!

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/saml2/Saml2LoginConfigurer.java [hunk 1]
```diff
@@ -211,7 +225,10 @@ public void configure(B http) throws Exception {
 	}
 
 	private AuthenticationProvider getAuthenticationProvider() {
-		AuthenticationProvider provider = new OpenSamlAuthenticationProvider();
+		OpenSamlAuthenticationProvider provider = new OpenSamlAuthenticationProvider();
+		if (this.userDetailsService != null) {
+			provider.setUserDetailsService(this.userDetailsService);
+		}
```
Comment (fhanik) [inline config/src/main/java/org/springframework/security/config/annotation/web/configurers/saml2/Saml2LoginConfigurer.java]: I like that you can explicitly set it. In addition to this, what if there is a 
```
@Bean 
UserDetailsService
``` 
in the context. We should pick up that bean too

File: samples/javaconfig/saml2login/src/main/java/org/springframework/security/samples/config/SecurityConfig.java [hunk 2]
```diff
@@ -40,27 +45,43 @@
 public class SecurityConfig extends WebSecurityConfigurerAdapter {
 
 	RelyingPartyRegistration getSaml2AuthenticationConfiguration() throws Exception {
-		//remote IDP entity ID
+		// remote IDP entity ID
```
Comment (fhanik) [inline samples/javaconfig/saml2login/src/main/java/org/springframework/security/samples/config/SecurityConfig.java]: Nothing wrong with reformatting as part of a PR. 
Would you be able to put that into a separate commit, one that doesn't change functionality, so that reviewers can focus on what actually changed?

File: samples/javaconfig/saml2login/src/main/java/org/springframework/security/samples/config/SecurityConfig.java [hunk 3]
```diff
@@ -40,27 +45,43 @@
...
+				.remoteIdpEntityId(idpEntityId).idpWebSsoUrl(webSsoEndpoint)
 				.credentials(c -> c.add(signingCredential))
 				.credentials(c -> c.add(idpVerificationCertificate))
 				.localEntityIdTemplate(localEntityIdTemplate)
-				.assertionConsumerServiceUrlTemplate(acsUrlTemplate)
-				.build();
+				.assertionConsumerServiceUrlTemplate(acsUrlTemplate).build();
+	}
+
+	private UserDetailsService userDetailsServiceMockup() {
```
Comment (fhanik) [inline samples/javaconfig/saml2login/src/main/java/org/springframework/security/samples/config/SecurityConfig.java]: We'd like to keep the samples to bare minimum.
I will work with you to find a better place for this configuration and how to test it (I'm working on that right now)

File: config/src/main/java/org/springframework/security/config/annotation/web/configurers/saml2/Saml2LoginConfigurer.java (hunk 1 above)
Comment (herrminni) [reply config/src/main/java/org/springframework/security/config/annotation/web/configurers/saml2/Saml2LoginConfigurer.java]: Hello Fillip,
thank you for your review!

I like your idea of automatically using the UserDetailsService from @Bean registry, if this is present.

However my implementation of manually setting (or maybe overriding) to a specific UserDetailsService was explicitly intended by me:
This way you would be able to have two different UserDetailsServices at the same time and handle SAML authentication in a different way. For example, you could explicitly create a new User as a sideeffect, if that one was not known in a local database. The other UserDetailsService could be used to only authenticate already known users an authenticate by username+password.

Maybe we could combine your and my idea:
User the spring @Bean UserDetailsService as default and also be able to overrride it with the method I already included (setUserDetailsService)?

File: samples/javaconfig/saml2login/src/main/java/org/springframework/security/samples/config/SecurityConfig.java (hunk 2 above)
Comment (herrminni) [reply samples/javaconfig/saml2login/src/main/java/org/springframework/security/samples/config/SecurityConfig.java]: Hi Fillip,
I used the suggested formatter config for Eclipse, which is part of spring-security from the file `etc/eclipse/eclipse-code-formatter.xml`.
Reverting the automatic formatting for the files I changed would have been too much manual work, and I had many local test-builds that failed because of the formatting. So I decided to commit the files with all the automatic changes from the eclipse-code-formatter that went through the test.

File: samples/javaconfig/saml2login/src/main/java/org/springframework/security/samples/config/SecurityConfig.java (hunk 3 above)
Comment (herrminni) [reply samples/javaconfig/saml2login/src/main/java/org/springframework/security/samples/config/SecurityConfig.java]: Sure, it just felt right for me, to add a sample.
Feel free to create a second example, that just uses the UserDetailService.
Let me know, if I can help!

**Comment by mballoni** [comment]
Wow I was looking exactly for it!
In case it gets merged soon, at which version it would be available ?

**Comment by mballoni** [comment]
@herrminni are you still planning on working on this PR?
I can pick up from were you left if you prefer.

**Comment by herrminni** [comment]
> @herrminni are you still planning on working on this PR?
> I can pick up from were you left if you prefer.

@mballoni As you can see, I replied to the review comments of @fhanik 
I'm not sure, how the review process will go on now.
Please feel free to add/change the PR yourself. I am happy to support you and Fillip on this

**Comment by mballoni** [comment]
@herrminni Thank you for pointing it out!

@fhanik fixing the merge conflicts would be enough for this PR or should we look deeper into some of those feedbacks?

**Comment by fhanik** [comment]
I think we'll be able to simplify this PR a lot after these two get merged

1. https://github.com/spring-projects/spring-security/pull/7693
2. https://github.com/spring-projects/spring-security/pull/7683

Those two PRs will get us

1. A testing framework for testing changes to Saml2LoginConfigurer
2. A `Saml2AuthenticatedPrincipal` interface that we will be able to extend
3. The ability to configure the authentication manager via a resolver

Once those two get through, we can rework this one, is that ok?

**Comment by mballoni** [comment]
Thank you for your quick response @fhanik .

Yes, that works for me.
Let me know if I can help with something.

I appreciate.

**Comment by mballoni** [comment]
@fhanik hi! 
Now that both of those issues were merged can we continue with this one?
Can I pick up this PR and take core of the merges or should I wait/do something else?

Thank you, I appreciate your patience.

**Comment by fhanik** [comment]
hi @mballoni - yes, please pick it up. I'll work with you on the review process.

**Comment by DEEPAKRDEEPS** [comment]
When you are planning to release this one ?

**Comment by DEEPAKRDEEPS** [comment]
Can I know when this will be get released?

**Comment by fhanik** [comment]
@mballoni If you're still interested, take a look at the Oauth2Login implementation.

The [user details get loaded here[1]](https://github.com/spring-projects/spring-security/blob/3b89754926b1322fc606ecfe5652dc455ea594fc/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/authentication/OAuth2LoginAuthenticationProvider.java#L116-L117)

Creating an implementation that makes similar choices to this implementation will make the code base easy to follow.

I'm closing this PR as the code has diverted away from this. We can attach the work to:
https://github.com/spring-projects/spring-security/issues/8010


[1] https://github.com/spring-projects/spring-security/blob/3b89754926b1322fc606ecfe5652dc455ea594fc/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/authentication/OAuth2LoginAuthenticationProvider.java#L116-L117
